### DISCUSSION TITLE
Signin with OAuth server side

### DISCUSSION THREAD

**Discussion by simpel** [supabase/supabase#21684] [Questions]
Hello!

I ran `npx create-next-app -e with-supabase` and changed login to use OAuth instead but nothing works. I just changed the server action to what you see below. See the log below. Why is this not working?

```typescript
const signIn = async (formData: FormData) => {
    "use server";

    const supabase = createClient();

    const { error, data } = await supabase.auth.signInWithOAuth({
      provider: "google",
    });

    console.log({ data, error });

    if (error) {
      return redirect("/login?message=Could not authenticate user");
    }

    return redirect("/protected");
  };
``` 

```
{
  data: {
    provider: 'google',
    url: 'https://**********************.supabase.co/auth/v1/authorize?provider=google&code_challenge=*********************&code_challenge_method=s256'
  },
  error: null
}
```

**Comment by GaryAustin1** [comment]
Oauth does not return the session to your app.  It will redirect to siteURL or your redirectTo option if you provide it and start running your code from there.

https://supabase.com/docs/guides/auth/server-side/oauth-with-pkce-flow-for-ssr

**Reply by simpel** [reply]
Hi @GaryAustin1!

Thanks for the reply! Does that then mean that this is the correct way of doing it? It kind of works...?

```typescript
'use server'

import { redirect } from 'next/navigation'
import { createClient } from '../clients/supabase/server'

export async function signInAction() {
	const supabase = createClient()

	const { data, error } = await supabase.auth.signInWithOAuth({
		provider: 'google',
		options: {
			redirectTo: 'http://localhost:3000/auth/callback',
		},
	})

	if (data.url) redirect(data.url)

	redirect('/error')
}
```

**Reply by GaryAustin1** [reply]
I don't use OAuth, but I believe you need to not do the redirect at the end to data.url.  That is already happening behind the scenes.   Unless there is an error you just need your code to stop there.  The browser is going to be redirected to the 3rd party interface and then back to your redirect URL.

![image](https://github.com/supabase/supabase/assets/54564956/88610bd6-cbde-4463-8495-06759bc67d82)

**Reply by simpel** [reply]
Hi! Yea, thats the way I read it too, that it should just solve itself but nothing happens if I don't do the redirect so I guess the documentation is a bit unclear here....

**Reply by GaryAustin1** [reply]
This is for sure the way it works.  The call should go to the auth provider screen.  Then, after signin with that, it should call back your app at either siteUrl or your redirectTo if it matches exactly your dashboard redirect settings.  Your app should be called back without you doing the redirect yourself unless you are doing something wrong or need to handle other things. 

I won't be able to help on this as I don't use these environments, but you for sure don't want to redirect on your own for this.

An example of Supabae making the call in React.
https://github.com/supabase-community/auth-ui/blob/main/packages/react/src/components/Auth/interfaces/SocialAuth.tsx

**Reply by J0** [reply]
Hey!

Thanks @GaryAustin1 for the detailed answers as always and thinks @simpel for taking time to explain the issue at hand. Can I quickly check that `createClient` is using @supabase/ssr and not @supabase/supabase-js ?  The [vanilla library likely doesn't redirect automatically](
https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts#L1836) in a server side context as there's no window.

Lmk!

**Reply by simpel** [reply]
Hey!

Sorry for the slow reply. I'm still banging my head on this one so any help is still appreciated. Here's my code;

### The client

```typescript
import process from 'process'
import { createServerClient, type CookieOptions } from '@supabase/ssr'
import { cookies } from 'next/headers'
import { redirect } from 'next/navigation'
import { type Database } from '@/types/TDatabase'

export const redirectToPath = (path: string) => redirect(path)

export const createClient = () => {
  const cookieStore = cookies()

  return createServerClient<Database>(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,

    {
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value
        },

        set(name: string, value: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value, ...options })
          } catch {
            console.error(
              'Error setting cookie in createClient() function in src/utils/supabase/server.ts file.',
            )
          }
        },

        remove(name: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value: '', ...options })
          } catch {
            console.error(
              'Error removing cookie in createClient() function in src/utils/supabase/server.ts file.',
            )
          }
        },
      },
    },
  )
}

``` 

### My server actions

```typescript

'use server'

import { type Provider } from '@supabase/supabase-js'
import { getErrorRedirect, getURL } from '../helpers'
import { createClient } from '../supabase/server'

export const signOutAction = async (formData: FormData) => {
  const currentPath = formData.get('currentPath')?.toString() ?? ''

  const supabase = createClient()
  const { error } = await supabase.auth.signOut()

  if (error) {
    return getErrorRedirect(
      currentPath,
      'Hmm... Something went wrong.',
      'You could not be signed out.',
    )
  }

  return '/'
}

export const signInAction = async (formData: FormData) => {
  const provider: Provider = (formData.get('provider')?.toString() ??
    'github') as Provider

  const supabase = createClient()
  const redirectURL = getURL('/auth/callback')

  await supabase.auth.signInWithOAuth({
    provider,
    options: {
      redirectTo: redirectURL,
    },
  })
}

``` 

### The form

```typescript
'use client'

import { usePathname } from 'next/navigation'
import SignInButton from '../SignInButton/SignInButton'
import { signInAction } from '@/utils/auth/actions'

const SignInForm = () => {
  const currentPath = usePathname()

  return (
    <form action={signInAction}>
      <input type="hidden" name="currentPath" value={currentPath} />
      <SignInButton />
    </form>
  )
}

export default SignInForm
```

**Comment by jimmymjing** [comment]
same issue. is it solved by now?

**Comment by j4w8n** [comment]
@jimmymjing can you be more specific about your issue?

If the OAuth login method is called on the client side, then the redirect - to authenticate - is done automatically.

If called on the server side, then you have to redirect to the returned `data.url` yourself.

After authentication is successful, then the user is automatically redirected to the url passed to `redirectTo` - whether the original method call happened client side or server side. If `redirectTo` is not defined, then I believe it sends the user to your Site URL.
